_posts/2021-10-25-choosing-chunk-and-partition-sizes.md

```diff
@@ -0,0 +1,79 @@
+tags: [performance]
+theme: twitter
+---
+{% include JB/setup %}
+
+Confused about choosing [a good chunk size](https://docs.dask.org/en/latest/array-best-practices.html#select-a-good-chunk-size) for Dask arrays?
+
+Array chunks can't be too big (we'll run out of memory), or too small (the overhead introduced by Dask becomes overwhelming). So how can we get it right?
+
+It's a two step process:
+
+1. First, start by choosing a chunk size similar to data you know can be processed entirely within memory (i.e. without dask)
+2. Then, watch the dask dashboard task stream and worker memory plots, and adjust if needed.
+
+## What are Dask array chunks?
+
+Dask arrays are big structures, made out of many small chunks.
+Typically, each small chunk is an individual [numpy array](https://numpy.org/), and they are arranged together to make a much larger Dask array.
+
+<img src="https://raw.githubusercontent.com/dask/dask/ac01ddc9074365e40d888f80f5bcd955ba01e872/docs/source/images/dask-array-black-text.svg" alt="Diagram: Dask array chunks" width="400" height="300" />
+
+## Too small is a problem
+
+If array chunks are too small, it's inefficient. Why is this?
+
+Using Dask introduces some amount of overhead for each task in your computation.
+This overhead is the reason the Dask best practices advise you to [avoid too-large graphs](https://docs.dask.org/en/latest/best-practices.html#avoid-very-large-graphs).
+This is because if the amount of actual work done by each task is very tiny, then the percentage of overhead time vs useful work time is not good.
+
+It might be hard to understand this intuitively, so here's an analogy.
+
+Let's imagine we're building a house. It's a pretty big job, and if there were only one worker it would take much too long to build.
+So we have a team of workers and a site foreman. The site foreman is equivalent to the Dask scheduler: their job is to tell the workers what tasks they need to do.
+
+Say we have a big pile of bricks to build a wall, sitting in the corner of the building site.
+If the foreman (the Dask scheduler) tells workers to go and fetch a single brick at a time, then bring each one to where the wall is being built, you can see how this is going to be very slow and inefficient! The workers are spending most of their time moving between the wall and the pile of bricks. Much less time is going towards doing the actual work of mortaring bricks onto the wall.
+
+Instead, we can do this in a smarter way. The foreman (Dask scheduler) can tell the workers to go and bring one full wheelbarrow load of bricks back each time. Now workers are spending much less time moving between the wall and the pile of bricks, and the wall will be finished much quicker.
+
+## Too big is also a problem
+
+If the Dask array chunks are too big, this is also bad. Why?
+Chunks that are too large are bad because then you are likely to run out of working memory.
+
+When this happens, Dask will try to spill data to disk instead of crashing.
```

Comment:
I've tried to address this by:
1. Also mention that users may see out of memory errors if the chunk size is too large, and...
2. Clarifying that several causes can result in spilling to disk. I've tried to edit it to include your phrasing, so now it says "When too much data is loaded in memory on too few workers, Dask will try to spill data to disk instead of crashing."